src/Indigo/Contracts/HomebaseLite/Optimizer.hs

```diff
@@ -0,0 +1,222 @@
+
+flattenSeqLHS :: Rule -> Rule
+flattenSeqLHS toplevel = Rule \case
+  it@(Seq (Seq _ _) _) -> Just $ linearizeAndReapply toplevel it
+  _                    -> Nothing
+
+linearizeAndReapply :: Rule -> Instr inp out -> Instr inp out
+linearizeAndReapply restart = \case
+  Seq (Seq a b) c ->
+    applyOnce restart $ Seq a (linearizeAndReapply restart (Seq b c))
+
+  other -> applyOnce restart other
+
+applyOnce :: Rule -> Instr inp out -> Instr inp out
+applyOnce r i = maybe i id (unRule r $ i)
+
+optimizerConf :: OptimizerConf
+optimizerConf = def {
+  ocRuleset = ocRuleset def <> fmap fixpoint
+    (seqAdapter (flattenSeqLHS
+      `orSimpleRule` unrollDrops
+      `orSimpleRule` rollDup
+      `orSimpleRule` rollDip)
+    : replicate 10 mainStage
+    <>
+     [seqAdapter $ flattenSeqLHS
+      `orSimpleRule` simpleSynonyms
+      `orSimpleRule` rollDrops
+     ]
+    ) <> ocRuleset def
+  }
+  where
+    mainStage =
+      seqAdapter (flattenSeqLHS
+        `orSimpleRule` dupDipDrop
+        `orSimpleRule` pushDipDug
+        `orSimpleRule` dupDugDrop
+        `orSimpleRule` notIf
+        `orSimpleRule` dupDipDownstream
+        `orSimpleRule` digDug
+        `orSimpleRule` redundantIf
+        `orSimpleRule` nestedDip
+        `orSimpleRule` emptyDip
+        `orSimpleRule` dropFrame
+        ) `orSimpleRule` dropNop
+
+dropFrame :: Rule
+dropFrame = Rule \case
+  FrameInstr Proxy x :# xs -> Just . unsafeCoerce $ x :# unsafeCoerce xs
+  _ -> Nothing
+
+dropNop :: Rule
+dropNop = Rule \case
+  x :# Nop -> Just x
+  Nop :# xs -> Just xs
+  _ -> Nothing
+
+rollDip :: Rule
+rollDip = Rule \case
+  DIP x :# xs -> Just $ DIPN One x :# xs
+  _ -> Nothing
+
+redundantIf :: Rule
+redundantIf = Rule \case
+  IF x y :# xs
+    | x == y
+    -> Just $ DROP :# x :# xs
+  _ -> Nothing
+
+emptyDip :: Rule
+emptyDip = Rule \case
+  DIPN _ Nop :# xs -> Just xs
+  _ -> Nothing
+
+nestedDip :: Rule
+nestedDip = Rule \case
+  (DIPN (n :: PeanoNatural n) (DIPN One (x :: Instr s s')) :: Instr inp out) :# xs
+    | Refl :: IsLongerOrSameLength inp ('S n) :~: 'True <- unsafeCoerce Refl
+    , Refl :: Take ('S n) inp ++ s :~: inp <- unsafeCoerce Refl
+    , Refl :: Take ('S n) inp ++ s' :~: out <- unsafeCoerce Refl
+    -> Just $ DIPN (Succ n) x :# xs
+  (DIPN One (DIPN (n :: PeanoNatural n) (x :: Instr s s')) :: Instr inp out) :# xs
+    | Refl :: IsLongerOrSameLength inp ('S n) :~: 'True <- unsafeCoerce Refl
+    , Refl :: Take ('S n) inp ++ s :~: inp <- unsafeCoerce Refl
+    , Refl :: Take ('S n) inp ++ s' :~: out <- unsafeCoerce Refl
+    -> Just $ DIPN (Succ n) x :# xs
+  _ -> Nothing
+
+unrollDrops :: Rule
+unrollDrops = Rule \case
+  DROPN One :# xs -> Just . unsafeCoerce $ DROP :# xs
+  (DROPN (Succ (n :: PeanoNatural n) :: PeanoNatural m) :: Instr inp out) :# xs
+    | Refl :: IsLongerOrSameLength inp n :~: 'True <- unsafeCoerce Refl
+    , Refl :: Drop n inp :~: a ': out <- unsafeCoerce Refl
+    -> Just $ DROPN n :# DROP :# xs
+  _ -> Nothing
```

Comment:
Wait, why this too?

`DROPN` should be better than multiple `DROP`, in fact we have an issue in `morley` to do exactly the opposite of this rule (that doesn't seem to be implementable for the same constraint issues)

Reply:
Quick answer to most of those un-optimizations, they are done at the first optimization pass to simplify other rules (i.e. I don't have to worry about matching `DROP` and _also_ `DROPN (Succ m)`). They are reversed at the last step (that said, there are a couple bugs in this here implementation, fixed in #11)

Reply:
> that doesn't seem to be implementable

Also, on this point, it _should_ be implementable, actually, at least I don't at first blush see any roadblocks. But whether it's worth it is another question, the proof will end up being rather long-winded as far as I can tell. Perhaps we need to rethink this whole optimizing the typed representation business and instead convert to untyped (or better yet simplified intermediate because there are commands in the untyped representation that do the same thing), then typecheck back.

Reply:
> it should be implementable, actually, at least I don't at first blush see any roadblocks

oh yes, I misspoke there, it should indeed be possible, it just turned out to be complex/tedious to do so.

> Perhaps we need to rethink this whole optimizing the typed representation business and instead convert to untyped

well, that's a fair point, but the idea of course is that working with typed morley has the usual advantage of types-checking (when not using `unsafeCoerce` :sweat_smile:), albeit at a fairly high implementation cost

Reply:
on this topic, last time this was needed was in `nbit`, see the `optimizeDrops` defined there.

The solution picked in that case was to write the "expanded" rules to match on `DROP` sequences, which preserved the type-checking.

Even tho that's hacky as well, I think it would give us (or at least me) some more peace of mind to go with a similar approach here, especially given that you noticed some bugs in this implementation. What do you think?

Reply:
> has the usual advantage of types-checking

Yes, but the cost is even with `unsafeCoerce` it's rather non-trivial to implement even the most obvious optimizations, like `DROP n; DROP m -> DROP (n+m)`, let alone anything more interesting. And there's actually a lot of interesting stuff one could do on the optimization side of things, but the compiler gets in the way in this case.

> write the "expanded" rules

:unamused: meh, that looks pretty horrible and not at all general. The bugs in this implementation don't have anything to do with types not aligning -- the only real issue is that `DROPN; DROP` rule's left side has incorrectly placed parentheses, hence it never matches. Less of an issue, but still unfortunate, is that `DROP; DROPN` rule is missing, which also comes in handy, it turns out. Nbit's solution doesn't solve either of these issues, and has a few obvious ones of its own.

If you're that concerned about type safety, I'd rather spend a few hours writing proofs to convince GHC that drop rules are ok (although morley doesn't export a way to get at `PeanoNatural` singletons, which will be rather annoying here). That said, network tests offer a reasonable guarantee the optimization rules didn't mess up anything particularly badly, and the only real concern is that those aren't applied when they should be -- type-safety won't help us with that.

Reply:
Alright, I agree that it's worth re-considering the use of typed morley in the optimizer rules.

I'm ok with your current definitions here for now, but that this should be fixed upstream in `morley`, even if we go with the
> spend a few hours writing proofs to convince GHC that drop rules are ok 

So for the time being please add a TODO here for `morley/#663` and `morley/#299` (which arguably if solved could remove the definitions here) so that we keep track of this